cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/ExternalDatasets.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.common;
+
+import co.cask.cdap.api.Admin;
+import co.cask.cdap.api.data.batch.Input;
+import co.cask.cdap.api.data.batch.InputFormatProvider;
+import co.cask.cdap.api.data.batch.Output;
+import co.cask.cdap.api.data.batch.OutputFormatProvider;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import com.google.common.base.Throwables;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class is used to create external datasets to track external sources/sinks.
+ */
+public final class ExternalDatasets {
+
+  /**
+   * If the input is an external source then an external dataset is created for tracking purpose and returned.
+   * If the input is a regular dataset or a stream then it is already trackable, hence same input is returned.
+   *
+   * @param admin {@link Admin} used to create external dataset
+   * @param input input to be tracked
+   * @return an external dataset if input is an external source, otherwise the same input that is passed-in is returned
+   */
+  public static Input makeTrackable(Admin admin, Input input) {
+    // If input is not an external source, return the same input as it can be tracked by itself.
+    if (!(input instanceof Input.InputFormatProviderInput)) {
+      return input;
+    }
+
+    // Input is an external source, create an external dataset so that it can be tracked.
+    String inputName = input.getName();
+    InputFormatProvider inputFormatProvider = ((Input.InputFormatProviderInput) input).getInputFormatProvider();
+    Map<String, String> inputFormatConfiguration = inputFormatProvider.getInputFormatConfiguration();
+
+    // Input is a dataset that implements input format provider,
+    // this too can be tracked by itself without creating an external dataset
+    if (inputFormatProvider instanceof Dataset) {
+      return input;
+    }
+
+    try {
+      // Create an external dataset for the input format for lineage tracking
+      String datasetName = "ext-" + inputName;
+      Map<String, String> arguments = new HashMap<>();
+      arguments.put("input.format.class", inputFormatProvider.getInputFormatClassName());
+      arguments.putAll(inputFormatConfiguration);
+      if (!admin.datasetExists(datasetName)) {
```

Comment:
The exists check is only for creating the external dataset. If the external dataset already exists then it is used. The input arguments are passed as runtime arguments to the external dataset during its instantiation. 

In the above case when B is deployed it sees that ext-myTopic exists, and uses that with the runtime arguments of input format provider.